File: modules/apps/desktop/ui/src/finding.test.ts
/**
 * The rules the palette is filled by, without a vault and without a browser.
 *
 * Three questions are in the air at once and come back in whatever order they
 * take, so the ones that matter here are the negatives: an answer to a question
 * nobody is asking any more must not fill a list a person is reading, and a
 * band that failed must not take the other two down with it.
 */
import { describe, expect, it } from 'vitest'
import { finding, type Asking, type Way, type Named, type Passage, type Words } from './finding'

const WORDS: Words = {
  names: 'Names',
  text: 'Text',
  meaning: 'Meaning',
  travel: 'Show in plex',
  read: 'Open the note',
  readAt: 'Open at this heading',
  readDocument: 'Open the document',
  noneFound: 'Nothing',
  notAsked: 'The vault could not answer',
  wordsOnly: 'Searching by words only — no model set',
  notEmbedded: 'This vault has not been read for meaning yet',
}

/** An answer the test hands over when it chooses to. */
interface Later<T> {
  promise: Promise<T>
  answers: (value: T) => void
  fails: (why: string) => void
}

function later<T>(): Later<T> {
  let answers!: (value: T) => void
  let fails!: (why: unknown) => void
  const promise = new Promise<T>((resolve, reject) => {
    answers = resolve
    fails = reject
  })
  return { promise, answers, fails: (why: string) => fails(new Error(why)) }
}

/** A vault that answers when the test says so, and remembers what it was asked. */
function asking() {
  const names: Later<readonly Named[]>[] = []
  const searched: { way: Way; answer: Later<readonly Passage[]> }[] = []
  const queries: string[] = []

  const core: Asking = {
    names: (query) => {
      queries.push(query)
      const one = later<readonly Named[]>()
      names.push(one)
      return one.promise
    },
    search: (query, way) => {
      queries.push(query)
      const one = later<readonly Passage[]>()
      searched.push({ way, answer: one })
      return one.promise
    },
  }

  const way = (which: Way) => searched.find((one) => one.way === which)?.answer
  return { core, names, searched, queries, way }
}

/** Nothing waits in a test; the hold is a clock and the clock is handed in. */
const now = async () => {}

/** Let everything already resolved run before the assertion. */
const settled = async () => {
  for (let turn = 0; turn < 6; turn += 1) await Promise.resolve()
}

const named = (over: Partial<Named> = {}): Named => ({
  path: 'notes/entropy.md',
  title: 'Entropy',
  heading: '',
  line: -1,
  at: [{ from: 0, to: 3 }],
  ...over,
})

const passage = (over: Partial<Passage> = {}): Passage => ({
  path: 'notes/heat.md',
  title: 'Heat engines',
  text: 'no engine beats a reversible one',
  isNote: true,
  start: 0,
  length: 0,
  line: 0,
  at: [{ from: 3, to: 9 }],
  ...over,
})

/** The band under one identity, from what the palette is drawing now. */
const bandOf = (bands: readonly { id: string }[], id: string) =>
  bands.find((one) => one.id === id) as
    | { id: string; items: readonly { id: string; title: string }[]; working?: boolean; silence?: string }
    | undefined

describe('asking', () => {
  it('draws no band at all until something is typed', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)

    expect(palette.bands.value).toHaveLength(0)
    expect(vault.queries).toHaveLength(0)
  })

  it('asks all three questions at once, about what was typed', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)

    void palette.typing('ent')
    await settled()

    expect(vault.names).toHaveLength(1)
    expect(vault.searched.map((one) => one.way)).toEqual(['words', 'meaning'])
    expect(new Set(vault.queries)).toEqual(new Set(['ent']))
  })

  it('asks about the words and not about the spaces around them', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)

    void palette.typing('  ent  ')
    await settled()

    expect(vault.queries).toEqual(['ent', 'ent', 'ent'])
  })

  it('asks nothing at all once what was typed is taken back', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)

    void palette.typing('ent')
    await settled()
    void palette.typing('')
    await settled()

    expect(vault.names).toHaveLength(1)
    expect(palette.bands.value).toHaveLength(0)
  })
})

describe('answers arriving', () => {
  it('fills each band on its own, while the others are still out', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)

    void palette.typing('ent')
    await settled()

    vault.names[0]?.answers([named()])
    await settled()

    expect(bandOf(palette.bands.value, 'names')?.items).toHaveLength(1)
    expect(bandOf(palette.bands.value, 'names')?.working).toBe(false)
    expect(bandOf(palette.bands.value, 'text')?.working).toBe(true)
    expect(bandOf(palette.bands.value, 'meaning')?.working).toBe(true)
  })

  it('drops an answer to a question nobody is asking any more', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)

    void palette.typing('ent')
    await settled()
    void palette.typing('entr')
    await settled()

    // The first question answers late, and with something else entirely.
    vault.names[0]?.answers([named({ path: 'notes/stale.md', title: 'Stale' })])
    await settled()
    expect(bandOf(palette.bands.value, 'names')?.items).toHaveLength(0)

    vault.names[1]?.answers([named()])
    await settled()
    expect(bandOf(palette.bands.value, 'names')?.items?.[0]?.title).toBe('Entropy')
  })

  it('says a band could not be asked in the window’s own words, and fills the others', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)

    void palette.typing('ent')
    await settled()

    vault.way('meaning')?.fails('no model is set')
    vault.names[0]?.answers([named()])
    await settled()

    expect(bandOf(palette.bands.value, 'meaning')?.silence).toBe(WORDS.notAsked)
    expect(bandOf(palette.bands.value, 'meaning')?.working).toBe(false)
    expect(bandOf(palette.bands.value, 'names')?.items).toHaveLength(1)
  })

  it('says a band came back with nothing in the window’s own words', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)

    void palette.typing('ent')
    await settled()
    vault.names[0]?.answers([])
    await settled()

    expect(bandOf(palette.bands.value, 'names')?.silence).toBe('Nothing')
  })

  it('says the vault holds no vectors, where meaning came back with nothing', async () => {
    const vault = asking()
    const read = { chunks: 4, embedded: 0, embedding: true }
    const palette = finding(vault.core, WORDS, now, () => read)

    void palette.typing('ent')
    await settled()
    vault.way('meaning')?.answers([])
    vault.names[0]?.answers([])
    await settled()

    expect(bandOf(palette.bands.value, 'meaning')?.silence).toBe(WORDS.notEmbedded)
    // The other bands are asked of the words, which a vault holding no vector
    // still answers.
    expect(bandOf(palette.bands.value, 'names')?.silence).toBe(WORDS.noneFound)
  })

  it('says nothing reads the vault for meaning, where nothing is set to', async () => {
    const vault = asking()
    const read = { chunks: 4, embedded: 0, embedding: false }
    const palette = finding(vault.core, WORDS, now, () => read)

    void palette.typing('ent')
    await settled()
    vault.way('meaning')?.answers([])
    await settled()

    expect(bandOf(palette.bands.value, 'meaning')?.silence).toBe(WORDS.wordsOnly)
  })

  it('lets go of everything when the palette is put away', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)

    palette.shows(true)
    void palette.typing('ent')
    await settled()
    palette.shows(false)

    vault.names[0]?.answers([named()])
    await settled()

    expect(palette.open.value).toBe(false)
    expect(palette.typed.value).toBe('')
    expect(palette.bands.value).toHaveLength(0)
  })
})

describe('where a thing found takes the person', () => {
  const filled = async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)
    void palette.typing('ent')
    await settled()

    vault.names[0]?.answers([
      named(),
      named({ path: 'notes/carnot.md', title: 'The Carnot cycle', heading: 'Entropy here', line: 12 }),
    ])
    vault.way('words')?.answers([passage()])
    await settled()
    return palette
  }

  it('opens a note found by its own name in the plex, and its text on the other key', async () => {
    const palette = await filled()
    const item = bandOf(palette.bands.value, 'names')!.items[0]!

    expect(palette.chose(item.id, 'plex')).toEqual({
      at: 'plex',
      path: 'notes/entropy.md',
      title: 'Entropy',
    })
    expect(palette.chose(item.id, 'note')).toEqual({
      at: 'note',
      path: 'notes/entropy.md',
      title: 'Entropy',
    })
  })

  it('opens a note found by a heading inside it on the line that heading stands on', async () => {
    const palette = await filled()
    const item = bandOf(palette.bands.value, 'names')!.items[1]!

    expect(palette.chose(item.id, 'note')).toEqual({
      at: 'note',
      path: 'notes/carnot.md',
      title: 'The Carnot cycle',
      line: 12,
    })
    // The plex draws notes and not the lines inside them.
    expect(palette.chose(item.id, 'plex')).toEqual({
      at: 'plex',
      path: 'notes/carnot.md',
      title: 'The Carnot cycle',
    })
  })

  it('opens a passage as the note it was read out of, on the line it stands on', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)
    void palette.typing('ent')
    await settled()

    vault.names[0]?.answers([named()])
    vault.way('words')?.answers([passage({ line: 12 })])
    await settled()
    const item = bandOf(palette.bands.value, 'text')!.items[0]!

    expect(palette.chose(item.id, 'note')).toEqual({
      at: 'note',
      path: 'notes/heat.md',
      title: 'Heat engines',
      line: 12,
    })
  })

  it('takes nobody anywhere for an item it is not drawing', async () => {
    const palette = await filled()
    expect(palette.chose('notes/nowhere.md', 'plex')).toBeNull()
  })
})

describe('what a key reaches, per kind of thing found', () => {
  it('offers the plex first for a name and the note first for a place', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)
    void palette.typing('ent')
    await settled()

    vault.names[0]?.answers([named(), named({ heading: 'Entropy here', line: 12 })])
    vault.way('words')?.answers([passage()])
    await settled()

    const acts = (band: string, at: number) =>
      (
        palette.bands.value.find((one) => one.id === band)?.items[at]?.actions ?? []
      ).map((one) => one.id)

    expect(acts('names', 0)).toEqual(['plex', 'note'])
    expect(acts('names', 1)).toEqual(['note', 'plex'])
    expect(acts('text', 0)).toEqual(['note', 'plex'])
  })
})

describe('a passage from something that is not a note', () => {
  it('offers the document, opened where the words were found', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)
    void palette.typing('war')
    await settled()

    vault.way('words')?.answers([
      passage({
        path: 'library/mahabharata.epub',
        title: '',
        isNote: false,
        start: 40_512,
        length: 31,
      }),
    ])
    await settled()

    const item = bandOf(palette.bands.value, 'text')!.items[0]! as {
      id: string
      title: string
      actions?: readonly { id: string; text: string }[]
    }

    expect(item.title).toBe('library/mahabharata.epub')
    expect(item.actions?.map((one) => one.text)).toEqual([WORDS.readDocument])

    // The document opens where the words stand in its own text.
    expect(palette.chose(item.id, 'document')).toEqual({
      at: 'document',
      path: 'library/mahabharata.epub',
      title: '',
      start: 40_512,
      length: 31,
    })
    // It is neither a note nor a node, and neither is answered for.
    expect(palette.chose(item.id, 'note')).toBeNull()
    expect(palette.chose(item.id, 'plex')).toBeNull()
  })
})

describe('a band landing under the keyboard', () => {
  it('names a passage by where it stands in the vault, not by where it stands in the list', async () => {
    const vault = asking()
    const palette = finding(vault.core, WORDS, now)
    void palette.typing('war')
    await settled()

    vault.way('words')?.answers([passage({ path: 'notes/heat.md' })])
    await settled()
    const before = bandOf(palette.bands.value, 'text')!.items[0]!.id

    // The same passage, now second in its band because a better one arrived.
    void palette.typing('war ')
    await settled()
    vault.searched
      .filter((one) => one.way === 'words')
      .at(-1)
      ?.answer.answers([
        passage({ path: 'notes/fire.md', title: 'Fire' }),
        passage({ path: 'notes/heat.md' }),
      ])
    await settled()

    const after = bandOf(palette.bands.value, 'text')!.items[1]!.id
    expect(after).toBe(before)
  })
})

File: modules/apps/desktop/internal/adapter/webui/finding_test.go
package webui_test

import (
	"testing"

	"connectrpc.com/connect"

	v1 "github.com/jiva-studio/numen/modules/libs/protocol/gen/numen/v1"
)

// The two questions the palette asks, over the wire.
//
// What a name or a passage is belongs to the index and the use case, and is
// asked there. What is asked here is what only the handler decides: which half
// the client's word for it runs, how many answers a number gets, and what
// reaches the client about a source that is not a note.

func TestANoteIsFoundByName(t *testing.T) {
	client, _ := opened(t, map[string]string{
		"Entropy.md": "---\ntitle: Entropy\n---\n\n# Entropy\n\n## Heat and work\n\nA reversible engine.\n",
		"Engines.md": "---\ntitle: Engines\n---\n\n# Engines\n",
	})

	answer, err := client.Names(t.Context(), connect.NewRequest(&v1.NamesRequest{Query: "entro"}))
	if err != nil {
		t.Fatal(err)
	}

	found := answer.Msg.GetFound()
	if len(found) == 0 {
		t.Fatal("a word still being typed found no name")
	}
	first := found[0]
	if first.GetNote().GetTitle() != "Entropy" {
		t.Errorf("the first answer is %q, want the note called that", first.GetNote().GetTitle())
	}
	if first.GetHeading() != nil {
		t.Errorf("a title matched, and the answer carries heading %+v", first.GetHeading())
	}
	if len(first.GetAt()) == 0 {
		t.Error("the answer says nothing about where it matched")
	}
}

func TestAHeadingIsFoundWithTheLineItStandsOn(t *testing.T) {
	client, _ := opened(t, map[string]string{
		"Entropy.md": "---\ntitle: Entropy\n---\n\n# Entropy\n\n## Heat and work\n\nA reversible engine.\n",
	})

	answer, err := client.Names(t.Context(), connect.NewRequest(&v1.NamesRequest{Query: "heat"}))
	if err != nil {
		t.Fatal(err)
	}

	found := answer.Msg.GetFound()
	if len(found) != 1 {
		t.Fatalf("%d answers, want the one heading: %+v", len(found), found)
	}
	heading := found[0].GetHeading()
	if heading.GetText() != "Heat and work" {
		t.Errorf("the answer is heading %q", heading.GetText())
	}
	if heading.GetLine() < 0 {
		t.Errorf("the heading stands on line %d, which is no line", heading.GetLine())
	}
}

func TestEachWayIsAskedByItself(t *testing.T) {
	client, _ := opened(t, map[string]string{
		"Engines.md": "---\ntitle: Engines\n---\n\n# Engines\n\nNo engine beats a reversible engine.\n",
	})

	// No model is set here, so the meaning half has nothing to answer with and
	// the words half answers on its own. Which is which is the handler's to get
	// right: both halves answer with the same shape.
	words, err := client.Search(t.Context(), connect.NewRequest(&v1.SearchRequest{
		Query: "reversible", Way: v1.Way_WAY_WORDS,
	}))
	if err != nil {
		t.Fatal(err)
	}
	if len(words.Msg.GetFound()) == 0 {
		t.Fatal("the words half found nothing for a word that is written")
	}

	meaning, err := client.Search(t.Context(), connect.NewRequest(&v1.SearchRequest{
		Query: "reversible", Way: v1.Way_WAY_MEANING,
	}))
	if err != nil {
		t.Fatal(err)
	}
	if got := len(meaning.Msg.GetFound()); got != 0 {
		t.Errorf("the meaning half answered %d times with no model set", got)
	}
}

func TestAPassageSaysWhichNoteItCameOutOf(t *testing.T) {
	client, _ := opened(t, map[string]string{
		"Engines.md": "---\ntitle: Engines\n---\n\n# Engines\n\nNo engine beats a reversible engine.\n",
	})

	answer, err := client.Search(t.Context(), connect.NewRequest(&v1.SearchRequest{
		Query: "reversible", Way: v1.Way_WAY_WORDS,
	}))
	if err != nil {
		t.Fatal(err)
	}

	found := answer.Msg.GetFound()
	if len(found) == 0 {
		t.Fatal("nothing found")
	}
	first := found[0]
	if first.GetNote().GetTitle() != "Engines" {
		t.Errorf("the passage says it came out of %q", first.GetNote().GetTitle())
	}
	if first.GetPath() == "" {
		t.Error("the passage says nothing about where it came from")
	}
	if len(first.GetAt()) == 0 {
		t.Error("the passage says nothing about where the word typed stands in it")
	}
}

// A hit is a place in a source, and the client is told which place: it is what
// opens the source there, and nothing the client holds says it.
func TestAPassageSaysWhereInItsSourceItStands(t *testing.T) {
	client, _ := opened(t, map[string]string{
		"Engines.md": "---\ntitle: Engines\n---\n\n# Engines\n\nNo engine beats a reversible engine.\n",
	})

	answer, err := client.Search(t.Context(), connect.NewRequest(&v1.SearchRequest{
		Query: "reversible", Way: v1.Way_WAY_WORDS,
	}))
	if err != nil {
		t.Fatal(err)
	}

	found := answer.Msg.GetFound()
	if len(found) == 0 {
		t.Fatal("nothing found")
	}
	first := found[0]
	if first.GetLength() == 0 {
		t.Errorf("the passage stands over no text: %d to %d",
			first.GetStart(), first.GetStart()+first.GetLength())
	}
	if first.GetStart() < 0 {
		t.Errorf("the passage begins at %d", first.GetStart())
	}
}

func TestAnAnswerIsCutToWhatWasAskedFor(t *testing.T) {
	client, _ := opened(t, map[string]string{
		"a.md": "# Entropy one\n",
		"b.md": "# Entropy two\n",
		"c.md": "# Entropy three\n",
	})

	answer, err := client.Names(t.Context(), connect.NewRequest(&v1.NamesRequest{
		Query: "entropy", Limit: 2,
	}))
	if err != nil {
		t.Fatal(err)
	}
	if got := len(answer.Msg.GetFound()); got != 2 {
		t.Errorf("%d answers for a limit of 2", got)
	}
}

func TestNothingTypedIsAnsweredWithNothing(t *testing.T) {
	client, _ := opened(t, map[string]string{"a.md": "# Entropy\n"})

	for _, query := range []string{"", "   "} {
		answer, err := client.Search(t.Context(), connect.NewRequest(&v1.SearchRequest{Query: query}))
		if err != nil {
			t.Fatalf("%q: %v", query, err)
		}
		if got := len(answer.Msg.GetFound()); got != 0 {
			t.Errorf("%q was answered with %d passages", query, got)
		}
	}
}

// What a note is divided into, over the wire.

func TestANoteSaysWhatItIsDividedInto(t *testing.T) {
	client, _ := opened(t, map[string]string{
		"Entropy.md": "---\ntitle: Entropy\n---\n\n# Entropy\n\n## Heat and work\n\nA reversible engine.\n",
	})

	answer, err := client.Headings(t.Context(),
		connect.NewRequest(&v1.HeadingsRequest{Paths: []string{"Entropy.md"}}))
	if err != nil {
		t.Fatal(err)
	}

	found := answer.Msg.GetFound()
	if len(found) != 1 || found[0].GetPath() != "Entropy.md" {
		t.Fatalf("found = %+v, want the one note asked about", found)
	}

	headings := found[0].GetHeadings()
	if len(headings) != 2 {
		t.Fatalf("headings = %+v, want the two the note carries", headings)
	}
	if headings[0].GetText() != "Entropy" || headings[0].GetLevel() != 1 {
		t.Errorf("the first is %+v", headings[0])
	}
	if headings[1].GetText() != "Heat and work" || headings[1].GetLevel() != 2 {
		t.Errorf("the second is %+v", headings[1])
	}
	if !(headings[0].GetLine() < headings[1].GetLine()) {
		t.Errorf("lines are %d and %d, want them in the order they stand",
			headings[0].GetLine(), headings[1].GetLine())
	}
}

func TestAPathAskedTwiceIsAnsweredOnce(t *testing.T) {
	client, _ := opened(t, map[string]string{
		"Entropy.md": "---\ntitle: Entropy\n---\n\n# Entropy\n",
	})

	answer, err := client.Headings(t.Context(), connect.NewRequest(&v1.HeadingsRequest{
		Paths: []string{"Entropy.md", "Entropy.md", "Entropy.md"},
	}))
	if err != nil {
		t.Fatal(err)
	}

	if found := answer.Msg.GetFound(); len(found) != 1 {
		t.Errorf("found = %+v, want one entry for the one note", found)
	}
}

func TestANoteCarryingNoHeadingIsAbsent(t *testing.T) {
	client, _ := opened(t, map[string]string{
		"Plain.md": "---\ntitle: Plain\n---\n\nProse and nothing else.\n",
	})

	answer, err := client.Headings(t.Context(), connect.NewRequest(&v1.HeadingsRequest{
		Paths: []string{"Plain.md", "Gone.md"},
	}))
	if err != nil {
		t.Fatal(err)
	}

	if found := answer.Msg.GetFound(); len(found) != 0 {
		t.Errorf("found = %+v, want nothing for a note with no headings", found)
	}
}
